### PR TITLE
Correction of #445 attach-detach problem

### PR DESCRIPTION
**Pull request by newbrain**
This commit should solve issue #455, introduced by PR #445.

Main changes, all in stm32l4.c:
- Both RAM and Flash are added only in the attach function, not in the probe one.
- The memory map is freed with `target_mem_map_free()` before adding the new map.
- To avoid proliferation of `switch{}` statements and slightly reduce `if` complexity, a new info data structure has been introduced. This is also propaedeutic to adding new similar targets without large changes in the code.
- Minor editing

Tested with STM32L476 Nucleo, Windows 10, gdb 8.0.1, stlink platform.

Especially the third point, will allow me to finally add the WB series without the need of a separate probe function (and additional source files, let's rework it once again).
BTW, the RM has now been published: [STM32WB55xx RM](https://www.st.com/content/ccc/resource/technical/document/reference_manual/group0/83/cf/94/7a/35/a9/43/58/DM00318631/files/DM00318631.pdf/jcr:content/translations/en.DM00318631.pdf) and also the Cube tools are available.

### CODE REVIEW

**Comment by UweBonnes** [comment]
Can you please merge into one patch and diminish unrelated whitespace changes?

File: src/target/stm32l4.c [hunk 1]
```diff
@@ -141,61 +243,62 @@ static void stm32l4_add_flash(target *t,
...
-		break;
-	case ID_STM32G07:
-		idcodereg = STM32G0_DBGMCU_IDCODE_PHYS;
-		break;
-	}
+	/* Retrive chip information, no need to check return */
+	struct stm32l4_info const *chip = stm32l4_get_chip_info(t->idcode);
+
+
+	uint32_t idcodereg = (chip->family == FAM_STM32G0x)
```
Comment (UweBonnes) [inline src/target/stm32l4.c]: This looks strange! You check for G0 and select something else,

File: src/target/stm32l4.c [hunk 2]
```diff
@@ -141,61 +243,62 @@ static void stm32l4_add_flash(target *t,
...
+	} else {
+		target_add_ram(t, 0x10000000, chip->sram2 << 10);
+		/* All L4 beside L47 alias SRAM2 after SRAM1.*/
+		uint32_t ramsize = (t->idcode == ID_STM32L47)?
+			chip->sram1 : (chip->sram1 + chip->sram2 + chip->sram3);
+		target_add_ram(t, 0x20000000, ramsize << 10);
+	}
+
+	/* Add the flash to memory map. */
+	/* TODO: This should really be more data driven... */
```
Comment (UweBonnes) [inline src/target/stm32l4.c]: What do you mean bt more data driven?

File: src/target/stm32l4.c [hunk 3]
```diff
@@ -124,6 +124,108 @@ struct stm32l4_flash {
...
+	FAM_STM32L4Rx = 2,
+	FAM_STM32G0x = 3,
+	FAM_STM32WBxx = 4,
+};
+
+#define DUAL_BANK	0x80u
+#define RAM_COUNT_MSK	0x07u
+
+struct stm32l4_info {
+	char *designator;
```
Comment (UweBonnes) [inline src/target/stm32l4.c]: The designator has constant length. In my understanding, char designator[10] will save us a pointer for every entry. Will fill up flash fast...

**Comment by UweBonnes** [comment]
Please comment on my remarks.

File: src/target/stm32l4.c (hunk 1 above)
Comment (newbrain) [reply src/target/stm32l4.c]: Expletives deleted.
Definitely my fault. The true and false branch should be reversed.

File: src/target/stm32l4.c (hunk 2 above)
Comment (newbrain) [reply src/target/stm32l4.c]: I really don't like the jumble of nested ifs, I would prefer something that takes the right action with fewer conditions (at the expense of more data, yes).
Maybe my working with the maintenance of SW with a Mc Cabe complexity of 183 has tainted me...
I've seen other todos here and there, but if you think this is not worth the reminder, I'll remove it.

File: src/target/stm32l4.c (hunk 3 above)
Comment (newbrain) [reply src/target/stm32l4.c]: Yes, very good observation.
I don't think there'll be a designator longer than 9 char, so I agree.
More or less 3 bytes per entry saved, no change to the code.

**Comment by newbrain** [comment]
Shall I provide a new commit (squashed with the previous one, so to only show one)?

File: src/target/stm32l4.c (hunk 3 above)
Comment (newbrain) [reply src/target/stm32l4.c]: Note that my solution was already less expensive than the original (+data -code)!
```
Mode                LastWriteTime         Length Name
----                -------------         ------ ----
-a----       2019-03-03     14:42          72284 blackmagic.bin <- master branch
-a----       2019-03-03     14:40          72192 blackmagic.nb  <- newbrain, char*
-a----       2019-03-03     14:38          72228 blackmagic.old <- newbrain char[10]
```

**Comment by newbrain** [comment]
Well, I went ahead and did it.
Eager to hear more comments, if any!

**Comment by UweBonnes** [comment]
Perhaps explain after  the commit message what was wrong, something like: With separate attach function, construction of memory map must happen during attach.

**Pull request merged**